### ISSUE TITLE
Link to CUDNN not working anymore + LORA training issue + Torch 2

### ISSUE THREAD

**Issue by Rojinski** [bmaltais/kohya_ss#909]
Hello ! 

First of all, and once again, thank you for all your work and bringing to us this powerfull tool. 

I have a **NVidia RTX 3070i (8 vram) - recent i9 - 32 mn ram - Windows 11 - CUDA 12** and I have no skill in coding.  I'm a sound engineer and soundtracks / dark ambient composer... ;)

Now the "problem" :) After upgrading, I had the issue of "modules installed outside the venv were found" (but there was not), i've made a clean reinstall (deleted everything). First with torch 2.0 (I used to run the repo with it and everything was ok)...  But i had to download CUDNN again and **your link doesn't work anymore**.  Error 404 or so... 

So, I've download the last version for my **CUDA 12** (**8.9.1.23**) on the NVidia website and it was a pain to complete the registration.  It comes in a zip file named "cudnn-windows-x86_64-8.9.1.23_cuda12-archive.zip". I've unzipped it and renamed the folder "cudnn_windows" and placed it in the root folder of Kohya. Then I've run the command mentionned on your page:

.\venv\Scripts\activate
python .\tools\cudann_1.8_install.py

It worked but it installed the **CUDNN 8.6** it seems....  and not the 8.9.1 for CUDA 12. Anyway, after all that, I tried to train a LORA and it was impossible. I get a long list of error that I've not copied (I'm sorry). The training couldn't start. After another clean reinstallation, (with torch 2 as before)., same issue. So, for the third try, I've chosen torch 1. Always the same pb with the install of the new CUDNN but, well, it seems tp work now. Slowly (7 hours 42 minutes for 200 images at 20 steps each / 10 epochs) but it works. It used to work with Torch 2 tho....  I thought it was a good idea to report this. Hope it helps.

Best regards

Rojinski
![Kohya](https://github.com/bmaltais/kohya_ss/assets/122015375/00c13daa-afdd-4d06-8152-ed99bd9a720c)

**Comment by bmaltais** [comment]
Hummm, this missing file is weird... Might have been pulled from the source site. I will see if I can rehost it somewhere else. I should have a local copy.

**Comment by bmaltais** [comment]
I don't have a copy of the file either. I tought I did but apparently not... darn... now I will need to find a way to make it from scratch.

**Comment by FurkanGozukara** [comment]
I have uploaded cudnn 8.7.0

But problem is it is training but not learning anything atm

I wonder if related cudnn or my torch or the new script

Installed torch 1.13 and cudnn 8.7.0

Tried with and without xformers

Not learning anything but working perfectly fine very weird

**Comment by FurkanGozukara** [comment]
0 learning

tested with
adamw8bit
adamw
fp16
bf16
on and off xformers

```
cuDNN v8.7.0 For CUDA 11.x

https://huggingface.co/MonsterMMORPG/SECourses/resolve/main/cudnn_windows%20v8.7.0.zip

pip3 install torch==1.13.1 torchvision torchaudio --index-url https://download.pytorch.org/whl/cu117

pip install https://huggingface.co/MonsterMMORPG/SECourses/resolve/main/xformers-0.0.21.dev541-cp310-cp310-win_amd64.whl

```

![image](https://github.com/bmaltais/kohya_ss/assets/19240467/9ac3fabc-9a90-4c65-ae4b-4be371587d5b)

![image](https://github.com/bmaltais/kohya_ss/assets/19240467/99b25740-737d-4975-8d1c-3fb45fe9a9f3)

**Comment by bmaltais** [comment]
I have found the file and I am hosting it on github... hoping it will last: https://github.com/bmaltais/python-library/raw/main/cudnn_windows.zip

**Comment by Rojinski** [comment]
> I have found the file and I am hosting it on github... hoping it will last: https://github.com/bmaltais/python-library/raw/main/cudnn_windows.zip

Thank you! :)

**Comment by Rojinski** [comment]
Well, I've tried your new setup.bat that I think it's crystal clear.... Thanks a lot. 
But...

I've tried to: 
1. clean the venv
2. install with torch 2.0
3. install cuDNN
4. run the gui
5. train a lora....

But then, after it started to train, i get this (see file in attachement) in the cmd windows. It was like that the last time I've tried to use with torch 2.  I insist on the following fact: it used to work before with torch2 ;)

Thanks a lot for your help.
[Kohya torch 2 error.txt](https://github.com/bmaltais/kohya_ss/files/11651041/Kohya.torch.2.error.txt)

**Comment by Rojinski** [comment]
After a clean reinstall with torch 1, it's working.... Too bad i can't use torch 2 again....  But, hey, it's working. :)

**Comment by FurkanGozukara** [comment]
> After a clean reinstall with torch 1, it's working.... Too bad i can't use torch 2 again.... But, hey, it's working. :)

can you verify that it is learning?

**Comment by bmaltais** [comment]
It is also reporting about some modules outside the bench... One of those might be conflicting as they take precedence over the venv modules.

Can you try uninstalling local modules as instructed?

Maybe this will make torch2 work?

**Comment by Rojinski** [comment]
> > After a clean reinstall with torch 1, it's working.... Too bad i can't use torch 2 again.... But, hey, it's working. :)
> 
> can you verify that it is learning?

Excuse me, but what do you mean by "learning"?  The result is ok when i try the Lora in Automatic 1111.

**Comment by Rojinski** [comment]
> It is also reporting about some modules outside the bench... One of those might be conflicting as they take precedence over the venv modules.
> 
> Can you try uninstalling local modules as instructed?
> 
> Maybe this will make torch2 work?

Once again, excuse me.... :) But what do i have to uninstall? What are the local modules you mention. Sorry for my ignorance... :/ I'm using other SD gui like Vlad, InvokeAI and Automatic 1111. They are all working fine with torch 2...

**Comment by bmaltais** [comment]
kohya_ss is not like vlad, invokaai or auto1111. It is mostly a trainer and use torch in a different manner and use other python requirements modules... so it will not necessarally work the same. It is possible your card will work to render prompt but not to train.

You also have multiple GPUs... so this might confuse kohya_ss. I have never tested the GUI with multiple GPUs.

**Comment by Rojinski** [comment]
> kohya_ss is not like vlad, invokaai or auto1111. It is mostly a trainer and use torch in a different manner and use other python requirements modules... so it will not necessarally work the same. It is possible your card will work to render prompt but not to train.
> 
> You also have multiple GPUs... so this might confuse kohya_ss. I have never tested the GUI with multiple GPUs.

As I told you earlier,  it was working, until a certain point, with torch 2.0. Then I've updated Kohya and it was gone.... I don't have multiple GPU, just one RTX 3070 ti. You make a confusion with  @FurkanGozukara message. I train embeddings on Automatic 1111 (torch 2.0) with no problem at all. And I loved (and still love it) kohya_ss for LORAs.  

Now it works perfectly under torch 1....  (768x768, 200 pictures, etc) But I wonder why I could train LORAs before with Kohya_ss under torch 2 and not now anymore.  So, i think it's not something on my side as I didn't change anything at all in my config. All my drivers are up-to-date, etc, etc....  I'd like to add that, before your update, i didn't get the message about "modules outside the venv"....  I just wanted to lnow what it means "outside the venv".... Does it mean "somewhere in my computer in other venv or used by other progs" or "somewhere in the kohya_ss root folder"?  It's not clear.  And I can see a lot of people have the same issue since recently....  even with more powerful GCs than my 3070ti. 

Don't take my words for an offense or so because it's absolutely not the case. I'm so grateful for this tool offered to the community *for free* ! I just try to understand.... Old habit of an old sound engineer. ;) And everything works faster with torch 2 (and it used to work before....).

![Kohya2](https://github.com/bmaltais/kohya_ss/assets/122015375/434491e9-b826-4a24-9492-e5b3a11a7455)

**Comment by Rojinski** [comment]
OK.... :) I think i have found the problem.... To cut a long story short:

**1. I've made another clean install (on a slow external usb HDD) tonight, one hour ago. I kept my torch 1 install on my SSD.
2. I've chosen torch2 (a lot of options  and choices to do have disappeared after a new upgrade, it seems)
3. I didn't install cuDNN manually this time!!!!!!!! But it seems it was installed anyway (cuDNN 8.7 --see the attached file) <<<-----  THAT'S THE POINT ! Maybe an update have been done to fix something?
4. I've started a LORA training of a precedant dataset (200 images, 768:768, 3 epochs, 20 steps)
5. IT WORKS and it's more than two times faster than with torch 1, cuDNN manually installed and xformers enabled... 1 hour instead of 2h42m !** 

**Conclusion:** something has been done or there's something wrong with installing cuDNN manually (as it seems it's installed automatically now) as indicated  here: https://github.com/bmaltais/kohya_ss#optional-cudnn-86...  I hope it helps....

Here's the copy of the CMD window. As you can see, it's still indicated the things about modules installed outside the venv.... But who cares? 

=============================================================
Modules installed outside the virtual environment were found.
This can cause issues. Please review the installed modules.

You can uninstall all local modules with:

deactivate
pip freeze > uninstall.txt
pip uninstall -y -r uninstall.txt
=============================================================

01:55:23-847878 INFO     nVidia toolkit detected
01:55:26-202895 INFO     Torch 2.0.1+cu118
01:55:26-216487 INFO     Torch backend: nVidia CUDA 11.8 cuDNN 8700
01:55:26-217488 INFO     Torch detected GPU: NVIDIA GeForce RTX 3070 Ti VRAM 8192 Arch (8, 6) Cores 48
01:55:26-218488 INFO     Verifying requirements
01:55:29-594438 INFO     headless: False
01:55:29-600445 INFO     Load CSS...
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
01:59:40-934884 INFO     Start training LoRA Standard ...
01:59:40-936894 INFO     Folder 20_Roji_MiaM: 200 images found
01:59:40-937884 INFO     Folder 20_Roji_MiaM: 4000 steps
01:59:40-938884 INFO     Total steps: 4000
01:59:40-939883 INFO     Train batch size: 1
01:59:40-939883 INFO     Gradient accumulation steps: 1.0
01:59:40-940884 INFO     Epoch: 3
01:59:40-941883 INFO     Regulatization factor: 1
01:59:40-941883 INFO     max_train_steps (4000 / 1 / 1.0 * 3 * 1) = 12000
01:59:40-942883 INFO     stop_text_encoder_training = 0
01:59:40-943884 INFO     lr_warmup_steps = 1200
01:59:40-944884 INFO     accelerate launch --num_cpu_threads_per_process=2 "train_network.py" --enable_bucket
                         --pretrained_model_name_or_path="F:/Safety Data/Stable Diffusion
                         Safety/models/Stable-diffusion/v1-5-pruned.safetensors"
                         --train_data_dir="E:/Images/AI/Training/Characters/Roji_MiaM/images" --resolution=768,768
                         --output_dir="E:/Images/AI/Training/Characters/Roji_MiaM/model"
                         --logging_dir="E:/Images/AI/Training/Characters/Roji_MiaM/log" --network_alpha="1"
                         --save_model_as=safetensors --network_module=networks.lora --text_encoder_lr=5e-05
                         --unet_lr=0.0001 --network_dim=8 --output_name="Roji_MiaM3" --lr_scheduler_num_cycles="3"
                         --learning_rate="0.0001" --lr_scheduler="cosine" --lr_warmup_steps="1200"
                         --train_batch_size="1" --max_train_steps="12000" --save_every_n_epochs="1"
                         --mixed_precision="fp16" --save_precision="fp16" --caption_extension=".txt" --cache_latents
                         --optimizer_type="AdamW8bit" --max_data_loader_n_workers="0" --bucket_reso_steps=64 --xformers
                         --bucket_no_upscale
A matching Triton is not available, some optimizations will not be enabled.
Error caught was: No module named 'triton'
prepare tokenizer
Using DreamBooth method.
prepare images.
found directory E:\Images\AI\Training\Characters\Roji_MiaM\images\20_Roji_MiaM contains 200 image files
4000 train images with repeating.
0 reg images.
no regularization images / 正則化画像が見つかりませんでした
[Dataset 0]
  batch_size: 1
  resolution: (768, 768)
  enable_bucket: True
  min_bucket_reso: 256
  max_bucket_reso: 1024
  bucket_reso_steps: 64
  bucket_no_upscale: True

  [Subset 0 of Dataset 0]
    image_dir: "E:\Images\AI\Training\Characters\Roji_MiaM\images\20_Roji_MiaM"
    image_count: 200
    num_repeats: 20
    shuffle_caption: False
    keep_tokens: 0
    caption_dropout_rate: 0.0
    caption_dropout_every_n_epoches: 0
    caption_tag_dropout_rate: 0.0
    color_aug: False
    flip_aug: False
    face_crop_aug_range: None
    random_crop: False
    token_warmup_min: 1,
    token_warmup_step: 0,
    is_reg: False
    class_tokens: Roji_MiaM
    caption_extension: .txt


[Dataset 0]
loading image sizes.
100%|███████████████████████████████████████████████████████████████████████████████| 200/200 [00:00<00:00, 203.98it/s]
make buckets
min_bucket_reso and max_bucket_reso are ignored if bucket_no_upscale is set, because bucket reso is defined by image size automatically / bucket_no_upscaleが指定された場合は、bucketの解像度は画像サイズから自動計算されるため、min_bucket_resoとmax_bucket_resoは無視されます
number of images (including repeats) / 各bucketの画像枚数（繰り返し回数を含む）
bucket 0: resolution (512, 768), count: 2780
bucket 1: resolution (768, 512), count: 1220
mean ar error (without repeats): 0.0
preparing accelerator
F:\kohya_ss\venv\lib\site-packages\accelerate\accelerator.py:258: FutureWarning: `logging_dir` is deprecated and will be removed in version 0.18.0 of 🤗 Accelerate. Use `project_dir` instead.
  warnings.warn(
Using accelerator 0.15.0 or above.
loading model for process 0/1
load StableDiffusion checkpoint: F:/Safety Data/Stable Diffusion Safety/models/Stable-diffusion/v1-5-pruned.safetensors
F:\kohya_ss\venv\lib\site-packages\safetensors\torch.py:98: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  with safe_open(filename, framework="pt", device=device) as f:
loading u-net: <All keys matched successfully>
loading vae: <All keys matched successfully>
loading text encoder: <All keys matched successfully>
CrossAttention.forward has been replaced to enable xformers.
import network module: networks.lora
[Dataset 0]
caching latents.
100%|████████████████████████████████████████████████████████████████████████████████| 200/200 [00:18<00:00, 10.72it/s]
create LoRA network. base dim (rank): 8, alpha: 1.0
neuron dropout: p=None, rank dropout: p=None, module dropout: p=None
create LoRA for Text Encoder: 72 modules.
create LoRA for U-Net: 192 modules.
enable LoRA for text encoder
enable LoRA for U-Net
preparing optimizer, data loader etc.

===================================BUG REPORT===================================
Welcome to bitsandbytes. For bug reports, please submit your error trace to: https://github.com/TimDettmers/bitsandbytes/issues
For effortless bug reporting copy-paste your error into this form: https://docs.google.com/forms/d/e/1FAIpQLScPB8emS3Thkp66nvqwmjTEgxp8Y9ufuWTzFyr9kJ5AoI47dQ/viewform?usp=sf_link
================================================================================
CUDA SETUP: Loading binary F:\kohya_ss\venv\lib\site-packages\bitsandbytes\libbitsandbytes_cuda116.dll...
use 8-bit AdamW optimizer | {}
running training / 学習開始
  num train images * repeats / 学習画像の数×繰り返し回数: 4000
  num reg images / 正則化画像の数: 0
  num batches per epoch / 1epochのバッチ数: 4000
  num epochs / epoch数: 3
  batch size per device / バッチサイズ: 1
  gradient accumulation steps / 勾配を合計するステップ数 = 1
  total optimization steps / 学習ステップ数: 12000
steps:   0%|                                                                                 | 0/12000 [00:00<?, ?it/s]
epoch 1/3
steps:  33%|███████████████████▎                                      | 4000/12000 [20:35<41:10,  3.24it/s, loss=0.124]
saving checkpoint: E:/Images/AI/Training/Characters/Roji_MiaM/model\Roji_MiaM3-000001.safetensors

epoch 2/3
steps:  42%|████████████████████████▌                                 | 5090/12000 [26:33<36:02,  3.20it/s, loss=0.124]

**Comment by NeoAnthropocene** [comment]
> OK.... :) I think i have found the problem.... To cut a long story short:
> 
> **1. I've made another clean install (on a slow external usb HDD) tonight, one hour ago. I kept my torch 1 install on my SSD. 2. I've chosen torch2 (a lot of options and choices to do have disappeared after a new upgrade, it seems) 3. I didn't install cuDNN manually this time!!!!!!!! But it seems it was installed anyway (cuDNN 8.7 --see the attached file) <<<----- THAT'S THE POINT ! Maybe an update have been done to fix something? 4. I've started a LORA training of a precedant dataset (200 images, 768:768, 3 epochs, 20 steps) 5. IT WORKS and it's more than two times faster than with torch 1, cuDNN manually installed and xformers enabled... 1 hour instead of 2h42m !**
> 
> **Conclusion:** something has been done or there's something wrong with installing cuDNN manually (as it seems it's installed automatically now) as indicated here: https://github.com/bmaltais/kohya_ss#optional-cudnn-86... I hope it helps....
> 
> Here's the copy of the CMD window. As you can see, it's still indicated the things about modules installed outside the venv.... But who cares?
> 
> Friendly,
> 
> Rojinski


Thank you so much. It solved my problem too.
